RationaleMCP/0037/modelica-uris.md

```diff
@@ -0,0 +1,120 @@
+- A string literal (or a substring thereof) appearing in a Modelica class definition, and in a position where a Modelica URI is given special meaning.  This generally excludes Modelica string literals denoting normal Modelica `String` values — a Modelica tool does not need to keep track of the origin of all string values in the form of Modelica URIs in order to preserve the class tree context.  Instead, it is in the context of certain annotations that a string literal can be in a position where a Modelica URI is expected, as in the `href` of an `a` tag in the `Documentation` annotation.
+- The [`resolveURI` operator](resolve-uri.md), also introduced by this MCP.
+
+
+### Base proposal: No authority, slash-separation, and Modelica lookup
+
+These are the different ways of referencing a class, where the _host_, _fullclass_, and _relclass_ represent a slash-separated Modelica class reference (that is, Modelica identifiers, separated by the forward slash (`/`) charater):
+- _modelica:/fullclass_ — Class given by its fully qualified name.  It is an error if _fullclass_ does not refer to an existing class.
+  * Example: _modelica:/Modelica/Electrical/Analog_
+  * Example: _modelica:///Modelica/Electrical/Analog_ (empty _authority_; the _fullclass_ form is the only possible form of class reference when _authority_ is present)
+- _modelica:relclass_ — Class given by lookup of _relclass_ in the class tree.  Requires class tree context, and it is an error if _relclass_ doesn't resolve to an existing class.
+  * Example: _modelica:Examples_
+  * Example: _modelica:_ (empty _relclass_)
+  * Example: _modelica:?figure=Disturbances&plot=Wind_ (empty _relclass_)
+- _modelica:./relclass_ — Same as _modelica:./relclass_, possibly useful to add clarity.
+  * Example: _modelica:./Examples_
+  * Example: _modelica:._ (empty _relclass_)
+  * Wrong: _modelica:///._ (malformed absolute reference with _fullclass_ `/.`)
+  * Wrong: _modelica:///./Examples_ (malformed absolute reference with _fullclass_ `/./Examples`)
+- _modelica:~/relclass_ — Similar to the form above, but lookup of _relclass_ is made from the point of the nearest enclosing encapsulated class, or the current top level class in case there is no enclosing encapsulated class.  For the MSL, where class encapsultion is currently not used much at all, this provides a convenient way to access resources organized in a hierarchy which is separate from the package hierarchy.
+  * Example: _modelica:~/Icons_
+  * Example: _modelica:~?resource=images/logo.png_ (empty _relclass_)
+  * Example: _modelica:~/Documentation?view=info#introduction_
+
+The forms containing _relclass_ are referred to as the _lookup-based forms_ (of a Modelica URI).  The form with _fullclass_ is referred to as the _fully qualified form_ (of a Modelica URI).
+
+### Lookup-free variation: Staying in the class tree
```

Comment:
I'm not convinced about adding these operations on their own without adding some additional groundwork.
The problem is that when copying classes we either need to make resources absolute (defeating the purpose of this), or copy the resources as well (which leads to duplicated resources). Thus I would propose to not delay this part, and study how it is handled in other systems.

Reply:
To see the full potential and need for relative class references, one needs to look beyond external resources.  A prime example is to refer to one of the figures of the current model from the model's documentation.  The figure is stored with the model, and accessing it by a fully qualified path is extremely inconvenient and hard to maintain.

When to copy or reuse external resources is a modeling and tool consideration in my mind.  The user needs to put external resources at a suitable level in the package hierarchy with class copying and sharing in mind.  A tool should offer good options for what to do with external resources when a class is copied — sometimes the user may prefer cloning the resources, sometimes not, and a helpful tool can fix the broken relative Modelica URIs.  A simple tool may not offer anything fancy at all, relying on the modeler to organize external resources and the references to them in such a way that things don't break when a class is copied naively.

Reply:
> To see the full potential and need for relative class references, one needs to look beyond external resources. A prime example is to refer to one of the figures of the current model from the model's documentation. The figure is stored with the model, and accessing it by a fully qualified path is extremely inconvenient and hard to maintain.

But the external resources are generally not stored in that way. In MSL the images are stored in Resources/Images/ and 3d-images in Resources/Data/Shapes.

Having the external data together with the Modelica-file was tried early on, and it failed in several ways in practice.

Obviously we can figure out something better, but my proposal is to not introduce relative paths until we have figured that out.

Reply:
> But the external resources are generally not stored in that way. In MSL the images are stored in Resources/Images/ and 3d-images in Resources/Data/Shapes.

The form relative to the encapsulation border is meant to serve this use case:
- _modelica:~?resource=Data/Tables/test_v4.mat_ (when referenced from within `Modelica`)

> Having the external data together with the Modelica-file was tried early on, and it failed in several ways in practice.

I wasn't there at the time, but I can see that this requires both experienced modelers and intelligent tool support to be a feature that really shines.  Of course, allowing non-absolute class references doesn't force anybody to use them for external resources, and they probably won't get much use until there is intelligent support in tools.

> Obviously we can figure out something better, but my proposal is to not introduce relative paths until we have figured that out.

But we need them desperately for other things than the external resources…

Reply:
If we discuss relative path to refer to something other than external resources, e.g., having the documentation of model `M` referring to a component (`a`) in its diagram layer then I strongly believe that we shouldn't view it as class-relative - but as a _normal reference_ in some way, so that if we have:
```
model Z
  M m;
end Z;
```
and open the documentation for 'm' then the documentation should refer to `m.a`, and below for `z.m` it should show `z.m.a`:
```
model Z2
  Z z(m(redeclare A2 a));
end Z2;
```
Similarly if we refer to a parameter `a.p` from `M` it should in `z.m` refer to `z.m.a.p`.

And some of problems with storing external data with the Modelica-files that were found: it relies on the file-system and is thus not ideal when storing models in a data-base, and if you change how package is stored it looks odd (e.g., when encrypting packages), and it was a bit messy as you cannot easily separate the resources from the rest.

Reply:
> If we discuss relative path to refer to something other than external resources, e.g., having the documentation of model M referring to a component (a) in its diagram layer then I strongly believe that we shouldn't view it as class-relative - but as a normal reference in some way

Well, maybe we just need to agree on what that _normal reference_ should look like.  I agree with the examples you give, but to me they way the references are presented in the documentation is a tool concern; what we need to solve here is how to express the reference.

To start with — to me — referencing a component from the class documentation is the job of a Modelica URI in one form or another.  Then, I don't see the problem of thinking of it as the component `a` of the current class, just as I believe it could be useful to refer to a component (possibly with a hierarchical name) in some other class.

If we begin by just considering the case of referencing a component of the current class, what would be the problem with using a Modelica URI in the following form?
- _modelica:?component=a_

When the documentation of `M` refers to `a.p`, it would be:
- _modelica:?component=a.p_

When some other class wants to reference the same parameter, via its appearance in `M`, it would be:
- _modelica:/M?component=a.p_

Reply:
> And some of problems with storing external data with the Modelica-files that were found: it relies on the file-system and is thus not ideal when storing models in a data-base, and if you change how package is stored it looks odd (e.g., when encrypting packages), and it was a bit messy as you cannot easily separate the resources from the rest.

A key point of the current proposal is that the _resources.d_ really separates the concern of how the Modelica package is split in a directory hierarchy from the concern of storing external resources.  I don't see how it relies on a file system — each class has its own directory of resources, and when a file system is used, that directory has a given place relative to the package root directory.  When a file system isn't used, resources belonging to a class will have to be managed some other way.  With the current proposal, however, a non-file-system tool that imports a packaged stored in a file system will know exactly which part of the file system hierarchy that contains external resources, and also to which classes these resources belong.

I also believe that this helps in the use case of encrypting a package.  My interpretation is that you would represent the class sub-tree to be encrypted as a single file, and the question is what to do with the rest of the files in the corresponding sub-tree of the file system — one just need to keep any _resources.d_ directories.  Is that any worse than what one would get with the approach of having all resources associated with the top level package?

Again, what the various users (including MAP-Lib) actually do with it is another question.  For this MCP, I'd like us to focus on the hard technical constraints, so that if there aren't any, we can give the users the freedom of using the language support for associating external resources with any class.

Reply:
> To start with — to me — referencing a component from the class documentation is the job of a Modelica URI in one form or another. Then, I don't see the problem of thinking of it as the component a of the current class, just as I believe it could be useful to refer to a component (possibly with a hierarchical name) in some other class.

For URIs referencing named Modelica-elements there are three possibilities:
* Local elements - components (and class parameters).
* Global references to classes (and elements in them).
* Lookup of classes (and possibly elements in them); in most cases global references would be safer - but in Media-packages it makes sense to use a similar lookup as normal for classes.

To me it would be ideal if the syntax for this matched the existing lookup in Modelica as much as possible. Having a special "?component=a" seems odd, and it will be difficult to remember what rules apply to class parameters, local lookup etc.

Reply:
> When a file system isn't used, resources belonging to a class will have to be managed some other way. With the current proposal, however, a non-file-system tool that imports a packaged stored in a file system will know exactly which part of the file system hierarchy that contains external resources, and also to which classes these resources belong.

But not all external resources are associated with one class, and having a separate Resources-directory (as used in MSL and the default in the current specification) seems to simplify import - as the Modelica files are imported and then there is one directory of "files" that needs to be stored internally as binary unstructured information.

>I also believe that this helps in the use case of encrypting a package. My interpretation is that you would represent the class sub-tree to be encrypted as a single file, and the question is what to do with the rest of the files in the corresponding sub-tree of the file system — one just need to keep any resources.d directories. Is that any worse than what one would get with the approach of having all resources associated with the top level package?

Having a clear separation of original Modelica-contents, and Resources, seems safer for encrypted contents where you want to include the encrypted Modelica-contents (assumedly just one file for the entire package), and Resources and nothing else.

Reply:
> > When a file system isn't used, resources belonging to a class will have to be managed some other way. With the current proposal, however, a non-file-system tool that imports a packaged stored in a file system will know exactly which part of the file system hierarchy that contains external resources, and also to which classes these resources belong.
> 
> But not all external resources are associated with one class, and having a separate Resources-directory (as used in MSL and the default in the current specification) seems to simplify import - as the Modelica files are imported and then there is one directory of "files" that needs to be stored internally as binary unstructured information.

A non-file system tool will still just need to manage one directory of "files": simply scrape the file system hierarchy for _resources.d_, and put these in a hierarchy that will make `resolveURI` easy.  For example, one can copy the directory hierarchy, keeping only any _resources.d_ directories.  However, there is no need to keep any of the structure at all, as long as each class in the database knows where its resources are located (and `resolveURI` can locate these locations).

> > I also believe that this helps in the use case of encrypting a package. My interpretation is that you would represent the class sub-tree to be encrypted as a single file, and the question is what to do with the rest of the files in the corresponding sub-tree of the file system — one just need to keep any resources.d directories. Is that any worse than what one would get with the approach of having all resources associated with the top level package?
> 
> Having a clear separation of original Modelica-contents, and Resources, seems safer for encrypted contents where you want to include the encrypted Modelica-contents (assumedly just one file for the entire package), and Resources and nothing else.

The use of _resources.d_ makes the separation crystal clear.  It would really not be difficult to grab all the _resources.d_ directories and nothing else.  Note that it is perfectly possible to combine the use of a hierarchy of _resources.d_ directories with a single file for the Modelica contents of the entire package.

Reply:
> For URIs referencing named Modelica-elements there are three possibilities:
> 
> * Local elements - components (and class parameters).
> * Global references to classes (and elements in them).
> * Lookup of classes (and possibly elements in them); in most cases global references would be safer - but in Media-packages it makes sense to use a similar lookup as normal for classes.
> 
> To me it would be ideal if the syntax for this matched the existing lookup in Modelica as much as possible. Having a special "?component=a" seems odd, and it will be difficult to remember what rules apply to class parameters, local lookup etc.

While I see that we could define Modelica URIs leveraging the full power of Modelica lookup, I'm also a bit hesitant about bringing in the entire instantiation machinery for something where it might not be really needed, at least not to start with.

Regarding the syntax, my current position is that it is more important to stay close to the inherit structure of a general URI, than having something that looks similar to Modelica.  The general URI structure gives us a well known way to parse the URI, and the use of queries gives natural extensibility that is hard to match with a non-standard structure such as the old Modelica URIs.  For example, if we don't introduce lookup of classes now, I don't see that it would be strange to introduce a new query for that in the future.  What sort of syntax would you propose for this instead?

By _class parameter_, do you mean something like a package constant?
- _modelica:/Modelica/Constants?component=pi

Compared to normal Modelica lookup, this has the advantage of clearly being a reference to a component, instead of having to perform lookup to see that the `pi` in `Modelica.Constants.pi` isn't a type.  This is probably going to make it easier and much cheaper to construct the correct hyperlink to the documentation for the component `pi` in `Modelica.Constants`.

Reply:
> To me it would be ideal if the syntax for this matched the existing lookup in Modelica as much as possible. Having a special "?component=a" seems odd, and it will be difficult to remember what rules apply to class parameters, local lookup etc.

I should have mentioned that we should at least think about how to make best use of the fragment specifier.  In the simple case of only referring to a component of the referenced class, it would certainly look nice compared to using the `component` query:
- _modelica:/Modelica/Constants#pi_ (compared to: _modelica:/Modelica/Constants?component=pi_)

However, I'm not sure that this would serve us well in the end, considering the combination of referencing a component, and asking for a resource of that component.  Imagine asking for the icon of the component `foo.bar` of the current class:
- _modelica:?view=icon#foo.bar_

Here, one would need to first resolve the fragment specifier, and then apply the `view` query, even though the query goes before the fragment specifier.

I believe it would be better to always leave the interpretation of the fragment specifier until the query has been resolved, possibly leading to different interpretations depending on context.  For example, I think we all expect that the fragment specifier can be used to link to an `id` or `a.name` in the documentation:
- _modelica:/Modelica?view=info#ConnectorEquations_
- _modelica:?component=foo.bar&view=info#introduction_ (documentation for the component `foo.bar` in the current class, at the element _introduction_)

Right now, I can't think of any other obvious use of the fragment specifier, but it seems like a useful opportunity to be able to give it other meanings in other contexts.

Reply:
Maybe it is time to identify some subtopics and start some PRs to add some structure to the discussion…

Reply:
The suggested way of using the new URIs in annotations for figures now shows how the fragment specifier could work as a really nice way of referencing a `Plot` in the context of a given `Figure`.  I think this is just the result of having developed a better understanding of how the fragment specifier is meant to be used: the power of the fragment specifier is really in its application relative to a resource specified by a query.  Therefore, when allowing more things with `identifier` inside a `Figure` in the future, one should make sure that the identifiers are still unique within the `Figure`, as that will enable making good use of the fragment specifier.

Applying this reasoning to the references to components seem to indicate that the fragment specifier could work in some situations by simply thinking of the components as being resources identified by their component reference relative to a class.  On the other hand, it isn't going to work in all situations as, for example, the fragment specifier already has another use inside documentation, and then having a query alternative to use instead seems necessary.  Exactly what this boils down to seems to complicated to explore here.

Reply:
> A non-file system tool will still just need to manage one directory of "files": simply scrape the file system hierarchy for _resources.d_, and put these in a hierarchy that will make `resolveURI` easy. 

I hadn't realized that the directories would be renamed as well, as default from `Resources` to `resources.d`.
To me that this looks like an unnecessary stylistic change, and I think this proposal is getting too many of them.

We clearly need to change from "modelica://Library/Resources/a.png" to "modelica:/Library/Resources/a.png" due to that part of the URI now being case-insensitive, and better fragments could be an improvement.

However, it's not clear that we need all the additional changes and everything is mixed together so it's hard to see which changes and needed and which ones are nice to have, etc. I would thus believe that we take a step back and first do the necessary changes and then optionally add the other ones.

Reply:
> > A non-file system tool will still just need to manage one directory of "files": simply scrape the file system hierarchy for _resources.d_, and put these in a hierarchy that will make `resolveURI` easy.
> 
> I hadn't realized that the directories would be renamed as well, as default from `Resources` to `resources.d`.
> To me that this looks like an unnecessary stylistic change, and I think this proposal is getting too many of them.

It is not unnecessary and just stylistic:
- The name _Resources_ is bad because it interferes with the storage of a class named `Resources`.
- The directory name _Resources_ is not reserved for storage of external resources.
- External resources are not required to be stored in the directory _Resources_.
- The name _Resources_ is only standardized for external functions, not external resources in general.
- The current rule _The first part of the path shall not be the name of a class in the package given by the authority._ is a cumbersome way of making the Modelica URI of an external resource unique.

> We clearly need to change from "modelica://Library/Resources/a.png" to "modelica:/Library/Resources/a.png" due to that part of the URI now being case-insensitive, and better fragments could be an improvement.
> 
> However, it's not clear that we need all the additional changes and everything is mixed together so it's hard to see which changes are needed and which ones are nice to have, etc. I would thus believe that we take a step back and first do the necessary changes and then optionally add the other ones.

The reason to keep all of these related changes in one MCP is that we have a unique opportunity of making things better now that the new Modelica URIs will be distinguishable from the old ones.

The central part of the new Modelica URIs is that they try to really leverage the power of general URI structure.  Keeping references to external resources in the form _modelica:/Library/Resources/a.png_ would really get in the way of achieving this.

Note that a library would be able to keep its external resource in their current locations as long as they are intended to be accessed using the deprecate Modelica URIs.  When the library author is ready to take the step to the new Modelica URIs, moving all external resources into a _resources.d_ directory will be a trivial task.

If we were to remove the _resources.d_ part entirely from this MCP, I would like to do it in such a way that it doesn't destroy the other aspects of the MCP.  One way it could be done would be to use a different query than `resource` for accessing stuff stored in the old way:
- _modelica:/Modelica?filepath=Resources/a.png_

While possible, I don't like the idea of introducing a `filepath` query, as it would make the transition harder to an improved storage model; to get rid of external resources residing anywhere in a library's directory structure, we'd need to deprecate also the `filepath` query in addition to the old Modelica URI form, with each deprecation being a nuisance to library authors.